### ISSUE TITLE
[K3][gfx950] Isolate IRIS epoch-lifetime and grouped-MoE numerical fixes

### ISSUE THREAD

**Issue by raikonenfnu** [raikonenfnu/tokenspeed#46]
## Summary

This issue isolates the two kernel-level defects found while investigating Kimi K3 AIME26 score variance on 8x MI350X (`gfx950`). The proposed branch contains only these fixes and their regression tests:

- Branch: [`raikonen/fix_k3_numerical_kernels`](https://github.com/raikonenfnu/tokenspeed/tree/raikonen/fix_k3_numerical_kernels)
- Commit: [`6bb88e4`](https://github.com/raikonenfnu/tokenspeed/commit/6bb88e4)

It intentionally contains **no** SMG seed transport, sampling, scheduler, prefix-cache reset, CI/eval YAML, or `force-deterministic-rsag` changes. Prefix caching, graph compilation, IRIS, and the Gluon kernels remain enabled.

### 1. IRIS reused a symmetric payload before every peer finished reading it

Code path: [`IrisAllReduce`](https://github.com/raikonenfnu/tokenspeed/blob/6bb88e4/tokenspeed-kernel/python/tokenspeed_kernel/ops/communication/iris.py#L345-L367) allocates one symmetric input buffer reused by successive collectives. The old protocol had only `ready_flags`:

1. rank writes epoch N payload;
2. rank publishes ready(N);
3. peers wait for ready(N), then read the payload;
4. a fast rank can return and write epoch N+1 into the same address while a slow peer is still reading N.

`ready` proves that production has happened; it does not prove that consumption has finished. Stream ordering on each rank is not a cross-rank lifetime guarantee. CUDA graph replay makes this easier to trigger because the same addresses and launch sequence are reused rapidly.

The same address was also loaded using `.cg`, although remote payload contents change in place between epochs. A cached line can therefore survive the flag transition and expose old data.

#### Fix

- Add per-block/per-rank [`done_flags`](https://github.com/raikonenfnu/tokenspeed/blob/6bb88e4/tokenspeed-kernel/python/tokenspeed_kernel/ops/communication/iris.py#L358-L367).
- Pass them through both public launch paths: [`all_reduce`](https://github.com/raikonenfnu/tokenspeed/blob/6bb88e4/tokenspeed-kernel/python/tokenspeed_kernel/ops/communication/iris.py#L407-L420) and [`all_reduce_two`](https://github.com/raikonenfnu/tokenspeed/blob/6bb88e4/tokenspeed-kernel/python/tokenspeed_kernel/ops/communication/iris.py#L475-L514).
- After accumulating all remote payloads, publish done(N) and wait for every peer's done(N) before allowing the next stream-ordered reuse. See the Triton [producer/consumer completion handshake](https://github.com/raikonenfnu/tokenspeed/blob/6bb88e4/tokenspeed-kernel/python/tokenspeed_kernel/ops/communication/iris.py#L576-L590) and the specialized Gluon [equivalent](https://github.com/raikonenfnu/tokenspeed/blob/6bb88e4/tokenspeed-kernel/python/tokenspeed_kernel/ops/communication/iris.py#L819-L837).
- Change reusable remote payload loads to cache-coherent `.cv`, e.g. [single-tensor payload loads](https://github.com/raikonenfnu/tokenspeed/blob/6bb88e4/tokenspeed-kernel/python/tokenspeed_kernel/ops/communication/iris.py#L560-L574) and [K3's packed two-tensor loads](https://github.com/raikonenfnu/tokenspeed/blob/6bb88e4/tokenspeed-kernel/python/tokenspeed_kernel/ops/communication/iris.py#L786-L817).

This is a synchronization/correctness problem, not merely “same seed gives different random samples.” Its possible symptoms include a hang, stale payload, or an epoch N/N+1 mixture in activations.

### 2. gfx950 grouped A16W4 SiTU decode combined routes with unordered atomics

The old `M <= 16` fast path made multiple route CTAs atomically add FP32 results into the same token row. Atomicity prevents lost updates, but it does not impose an addition order. Floating-point addition is not associative, so CTA arrival order changes the rounded sum. That perturbation propagates into later layers/logits and can change a near-tie token even with identical inputs and seed.

#### Fix

The grouped path now emits one race-free BF16 partial per `(token, route)`, then invokes the existing masked reduction in fixed route order. The complete selection and reduction path is at [`gluon_a16w4_situ_grouped.py:811-880`](https://github.com/raikonenfnu/tokenspeed/blob/6bb88e4/tokenspeed-kernel-amd/python/tokenspeed_kernel_amd/ops/moe/gluon_a16w4_situ_grouped.py#L811-L880). Tiny `M <= 4` decode was already dispatched to the deterministic route-direct kernel and is unchanged.

This is confirmed numerical nondeterminism: before the fix, identical M=16 inputs changed 7 BF16 elements over 100 launches, with maximum absolute delta 0.25. M=5 and M=8 happened to be exact in that sample, which illustrates why an occasional passing decode size does not make the atomic algorithm deterministic. After the fix, M=5/8/16 are all bit-exact over 100 repeated launches.

## Reproduction and validation

Hardware: 8x AMD Instinct MI350X, gfx950.

### IRIS production-shape graph stress

The regression test alternates `all_reduce(16x7168)` and `all_reduce_two(16x7168, 16x3584)` through the same state, then captures both in one CUDA graph and replays it. It also repeats the fused all-reduce + residual + RMSNorm K3 attention path. See [`test_iris_communication.py:207-350`](https://github.com/raikonenfnu/tokenspeed/blob/6bb88e4/tokenspeed-kernel/test/ops/test_iris_communication.py#L207-L350).

On the unfixed commit, the composite alternating graph hung. A lone-collective graph did not reproduce the bug. On this branch, the world-8 test passed in 8.05 s, including 64 eager alternating epochs, graph replays, and fused-path eager/graph checks against exact BF16-rounded references.

```bash
pytest -q -s tokenspeed-kernel/test/ops/test_iris_communication.py::test_iris_all_reduce_correctness_world8
# 1 passed, 1 warning in 8.05s
```

### Gluon grouped-MoE repeatability and latency

The regression test uses fixed weights/routes/inputs and compares 100 launches exactly for M=5, 8, and 16. See [`test_gluon_a16w4_situ_grouped_gfx950.py`](https://github.com/raikonenfnu/tokenspeed/blob/6bb88e4/tokenspeed-kernel-amd/test/ops/test_gluon_a16w4_situ_grouped_gfx950.py#L23-L80).

```bash
pytest -q -s tokenspeed-kernel-amd/test/ops/test_gluon_a16w4_situ_grouped_gfx950.py
# 3 passed in 2.87s
```

A 25-warmup/200-launch synthetic decode microbenchmark (48 local experts, top-k 8, hidden/intermediate 256) gave:

| M | exact across 200 | median | p95 | min |
|---:|:---:|---:|---:|---:|
| 5 | yes | 120.5 us | 136.2 us | 116.6 us |
| 8 | yes | 119.6 us | 129.8 us | 114.6 us |
| 16 | yes | 120.1 us | 130.6 us | 116.6 us |

Repository hooks also pass:

```bash
pre-commit run --all-files
# all applicable hooks passed
```

## Accuracy/score claim boundary

These fixes remove two mechanisms that can corrupt or perturb K3 activations/logits, so they are correctness fixes worth landing independently of seed plumbing. They are **not proof** that either defect caused a particular AIME26 miss or that they alone guarantee a target score. Prior 0.8667/0.9000 AIME runs also contained 32768-token truncation loops, which are a separate evaluation-quality factor.

An end-to-end AIME26 comparison could not be completed on this node: checkpoint staging timed out and the retry approached filesystem capacity before the server became ready, so no inference request was issued. Only abandoned `.incomplete` downloads were removed; finalized checkpoint data was preserved. The kernel-level value and latency validations above did complete successfully.

Once a complete checkpoint is available, the follow-up acceptance run should use the current graph-enabled, IRIS-enabled, Gluon-enabled, prefix-cache-enabled eval configuration and compare several full AIME26 runs—not require bit-identical text—to check score distribution and truncation rate.

**Comment by raikonenfnu** [comment]
### Unfixed-vs-fixed latency ablation

I reran the identical 25-warmup/200-launch synthetic benchmark against base `2fb7815c` so the correctness fix has a visible cost baseline.

| M | base median / p95 | base exact | fixed median / p95 | fixed exact |
|---:|---:|:---:|---:|:---:|
| 5 | 110.1 / 130.2 us | yes | 120.5 / 136.2 us | yes |
| 8 | 111.6 / 130.5 us | yes | 119.6 / 129.8 us | yes |
| 16 | 112.4 / 124.2 us | **no** | 120.1 / 130.6 us | **yes** |

The fixed route-partial reduction costs about 8–10 us (~7–9%) in this small synthetic microkernel while removing the reproduced M=16 numerical race. This is not an end-to-end K3 throughput measurement; model-level impact should be measured once the complete checkpoint is available.
